### ISSUE TITLE
Rendering bugs with the combo of select multiple + dispatch 'input' event

### ISSUE THREAD

**Issue by DaTrader** [phoenixframework/phoenix_live_view#2718]
### Environment

* Elixir version (elixir -v): 1.14.4
* Phoenix version (mix deps): 1.7.3
* Phoenix LiveView version (mix deps): 0.18.18
* Operating system: Linux Mint
* Browsers you attempted to reproduce this bug on (the more the merrier):  FF, Chrome
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no: Yes

### Actual behavior

The below is a description of the components and the flow producing the bug, although I can't tell if it would surface in a different setup.

There are two live components . One is a master form that collects various data (Master), The other is a specific detail selection component (Selector). On user request, Master opens Selector (the `:if` attribute on an assign). They are both modal components, where Master is hidden (but live) when Selector is open. When is done using Selector, it sends update to Master (Master having passed its id to Selector when opening it), Master takes the assigns, updates and renders its own including closing Selector.

The part that creates the bug is in Master where a hidden multiple `select` is used to render the received item's ids for the purpose of having them submitting along with the other Master form fields once the user clicks the submit button and for having the `phx-no-feedback` class automatically added to the relevant element. The Master relies on dispatching the 'input' event as explained in the LiveView docs (`https://hexdocs.pm/phoenix_live_view/form-bindings.html#triggering-phx-form-events-with-javascript`):

```
<.select
  field={@field} multiple={true} options={@selected} value={@selected} class="hidden"
  x-data="{ oldSelected: '' }"
  x-init={"
    selected = '#{ Phoenix.HTML.javascript_escape( Enum.join( @selected, ","))}';
    if( oldSelected !== selected) {
      oldSelected = selected;
      $el.dispatchEvent( new Event( 'input', { bubbles: true}));
    }
  "}
/>
```

The erroneous behavior is as follows. When Master first opens Selector and the user selects the items, Master renders the hidden options properly and it all seems to work. Now, when it opens Selector for a second  time - while still containing the old selection (say, to change the selected items), what happens 90% of the time is that apparently does not render anything new when received the selection (i.e. Master keeps displaying the old selection), but what happens under the surface is that every single option line DOES get processed and rendered server side, but they don't reach the client. I replaced the `select` function component `Phoenix.HTML.Form.options_for_select/2` call with an explicit `<option/>` element with a `:for` comprehension and logged the rendering iterations to the console. They all render, but they don't get to the client.

I believe I tried every conceivable combination of dispatching the event with a delay (`setTimeout`) or dispatching it from `x-effect` instead of `x-init`, dispatching it from a different element after the `select`, adding the `id`'s everywhere, etc., etc. The end result was always the same. Once the event is removed, it all works (except in that case the form does not receive the intended event, of course).

For pragmatical reasons, I ended up removing the whole `select` element thing and validating the form manually based on the assigns received from Selector and adding the `phx-no-feedback` class to the relevant element manually. However, that's not how this was supposed to work, right?

Now, after a weekend of thinking through, I tried something else to circumvent this bug and I it worked. The bug may be circumvented by replacing the template code above with the following:

```
<.select
  field={@field} multiple={true} options={@selected} value={@selected} class="hidden"
  x-ref="select"
/>

<div
  class="hidden"
  x-init={"
    triggerSelected = #{ @trigger_selected};

    if( triggerSelected) {
      $refs.select.dispatchEvent( new Event( 'input', { bubbles: true}));
      console.log( `dispatching ${ triggerSelected}`);
    }
  "}
></div>
```

And then in the Master component code, where `selected` is assigned, have the component send update to itself and then in that other update (therefore in a next rendering cycle) assign an increment of the `trigger_selected` counter assign (which is initial set to false to avoid `x-init` dispatching the event when first rendered).

Since the workaround is introducing new code, new state and an additional rendering cycle only to babysit the bug, I decided I'd stick to the manual form validation/phx-no-feedback adding until the bug is fixed.

### Expected behavior

In short, the `input` event when dispatched, it shouldn't mess up the order of rendering/updating the client relative to the component actual state.

**Comment by josevalim** [comment]
Hi @DaTrader, as usual, we will need a mechanism to reproduce this issue to be able to move forward. Thanks!

**Comment by DaTrader** [comment]
@josevalim I thought so and I added it to my free time todo list. In the meantime, maybe someone else who experiences the same or similar issue provides their.

**Comment by DaTrader** [comment]
@josevalim Maybe a subject for another issue, but this bug aside, I have a suggestion regarding cases like this design-wise. 

Namely, I have several situations where this approach works because there I don't use a `select`, but just a simple hidden `input` as just a single value is required. However, to be honest I am not happy with this approach as it forces the client side to be the source of truth even when it isn't. The official LV docs mention "javascript" components altering the field and that would be a client side source of truth, but here what I have are pure LiveComponents where any selected item data is passed via `send_update`.

Also, all such cases (when they do work) require 2 update cycles not one (naturally, because the first is the value update by the LiveView, and only the second is the reaction to the dispatched `input` event). And this is what I dislike - two over the wire update cycles instead of one.

I haven't had the time to analyze the JS part of the LiveView for the past couple of years but I imagine the entire `phx-no-feedback` thing is managed completely on the client-side. I have also noticed that adding the `phx-feedback-for` from Alpine doesn't work - i.e. the attribute gets conditionally added or removed by LiveView does not react to it in terms of managing the `phx-no-feedback` class.

In order to keep the update cycle count while at the same time not having to manually track (via assigns) the dirty/altered fields the state of which is managed in the component code, my suggestion is to have LiveView either detect a user Alpine/hook managed `phx-feedback-for` (when one is not provided in the template) or to introduce a mechanism where this flag is passed on from the server when the field value is first changed programmatically to a non-whatever-is-considered-empty value.

What do you say?

**Comment by DaTrader** [comment]
@josevalim No problem. 

In short, if LV client-side detected (as opposed to continued to ignore) the `phx-feedback-for` attribute set by Alpine or the hook code in general (when not in conflict with the same attribute being set in the template), then this bug could be classified as obsolete along with all other cases of dispatching an `input` event only to have LV recognize an input value change as result of its own rendering (plus some other positives).

Just sayin..

**Comment by josevalim** [comment]
Ping!

**Comment by DaTrader** [comment]
I think another person already had a similar issue with the <select> option rendering. It could be it's a different bug though. As stated in one of my comments above, I went on with another approach so I need to reconstruct this one.  I'll see what I can do this weekend.

**Comment by DaTrader** [comment]
@josevalim Closing this for I can't reproduce the bug anymore. Can't tell if it's a difference b/w 0.19.x and 0.18.8 or it's just because I am failing to implement the bug-relevant part of the circumstances. The demo app is simpler bar far than the real app I am working on but I did recreate the send_update based communication from the child component back to its master and the dispatchEvent in Alpine code as described in the issue but the multi select renders now properly each time. Kinda spent 4+ hrs in vein :( but at least the good news is that maybe the bug does is no more.

**Comment by DaTrader** [comment]
@josevalim Btw, congrats on the async functionality added in LV 0.20.0. Any chance for us mortals to get a glimpse on the enhancement (feature) plans somewhere? It's been a constant dilemma whether to make some missing feature on my own or wait for a new LV version. Would be of great help to be able to know the feature roadmap upfront.

**Comment by chrismccord** [comment]
I have detailed the roadmap in my last few conference talks so that remains the official reference. Async stuff came out “unexpectedly” from my product work at fly.

**Comment by DaTrader** [comment]
@chrismccord Understood. The async stuff is more than welcome. I'm more referring to things like streams (and the phx-mounted, phx-remove attributes). I needed something like that from day 1 on my first LiveView project (4 years ago) and by the time streams were out I developed several list-based diffing mechanisms, including the one for this last project I am working on now. If only I had known a year ago the streams would be added to LiveView, it would have saved me something like a month of work then.

**Comment by josevalim** [comment]
The issue is that we can't promise any deadlines. So even if 4 years ago we thought we'd eventually implement streams, it could have been in 6 months or 3 years, and I bet you probably could not wait for 3 years nor 6 months. So if you need it now, the best is for you to not depend on anyone. :) For what is currently on our mind, the keynotes cover them. :)

**Comment by DaTrader** [comment]
It's clear you can't promise any deadlines, but a priority list would have done as well. Surely, if streams had been planned 4 years ago and got priority 1 (with 10 being the top priority), no one in their right mind would wait it, but maybe then they (we) could have had an opportunity to vote certain features higher on the list (that is if you as the core team would allow for something like that).
